### ISSUE TITLE
why can't I reproduce results after restoring the best model?

### ISSUE THREAD

**Issue by YUIWK** [autonomio/talos#246]
**Hello, mikkokotila. I think I might face the same problem ever talked about in #187, that is I can't reproduce results that I got from the experiment after restoring the best model. I tried to reprodue results on the training data set, validation data set but I failed. I don't know what I should do now.**

**Here are the codes. If you need others please tell me** 
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### 
ta.Deploy(h,'model_1',metric = 'val_loss',asc=True)
model_1 = ta.Restore('model_1.zip')
Y_train_hat = model_1.model.predict(X_train) #I have splited data into the trainning, validation, test data set.
Y_val_hat = model_1.model.predict(X_val)
Y_test_hat = model_1.model.predict(X_test)

#I calculated the square of e_train, e_val, e_test but they aren't displayed correctly 
e_train = Y_train-Y_train_hat
mse_train = sum(e_train**2)/e_train.shape  #output 1.46 but should be 1.85
e_val = Y_val-Y_val_hat
mse_val = sum(e_val**2)/e_val.shape #output 2.89 but should be 2.18
e_test = Y_test-Y_test_hat
mse_test = sum(e_test**2)/e_test.shape 
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### 

**By the way, I've looked through #187 but I didn't understand the last paragraph you said. Could you clarify it again for me? Thank you for your help in advance!** 
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### 
The way scan_object.data and the accompanying .csv gets the results from metrics is through the history object that comes out from model.fit() in Keras. There are two modes in which Talos can do this:
* based on highest epoch (which is default)
* based on last epoch(you have to set` Scan( ... last_epoch_value=True ... )`)
As I have seen, in your model you are currently using `last_epoch_value=False` which means you are unlikely to ever get exactly the same result from evaluation as you will get from the Scan() results. I think based on this experience, we will have it True by default going forward. Still then you will end up with the problem, which has to do with Keras, where whatever you have as your last epoch metrics, is not going to be the same as the one you get from evaluate (or accuracy_score from sklearn). For example, I tried with your configuration and made the corresponding Keras model directly instead of using Talos, and the last epoch result is roughly 2% different from what both evaluate and accuracy_score gives.
As I now know what you are looking for and your data, I can confirm that this is not an issue in Talos, and even though it is something that is known about Keras (there are many issues discussing this), I'm not sure it can be considered an issue because the last epoch training result and evaluation are really not the same thing. **You get exactly the same level of reproducibility with Talos as you get with Keras. There is absolutely no difference.**
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###

**Comment by mikkokotila** [comment]
First point, as it comes out from #187 is that generally speaking reproducibility is tricky when it comes to deep learning. Second point is that Talos provides exactly the same level of reproducibility as Keras does, as Talos not change Keras in anyway. 

The root of your issue is that you are not declaring `ta.Scan(... last_epoch_value=True ...)` in your experiment. That means the `val_loss` in the results data from Scan is not the last epoch result, but the epoch with the highest value in the permutation. In other words, it might be that the model became worse towards the end. So if you want reproducibility, you have to first follow everything that you would have to follow with Keras to get reproducible results, and then second you have to `ta.Scan(... last_epoch_value=True ...)`. 

Closing here. Feel free to open new issues if new issues arise.

**Comment by YUIWK** [comment]
Thanks for your quick reply. I think it's clearer for me now. Regarding ''last_epoch_value=True'', I realized this may be one of two reasons that caused the reproducibility problem but still cann't grasp the meaning or the role of it. Sorry for being obtuse about it.
re: That means the `val_loss` in the results data from Scan is not the last epoch result, but the epoch with the highest value in the permutation
What do you mean by that? Especially, what the word epoch means? It's one of the hyperparameters that we define in the dictionary? I think maybe I miss some details of how Talos work.

**Comment by mikkokotila** [comment]
>>That means the `val_loss` in the results data from Scan is not the last epoch result, but the epoch with the highest value in the permutation.

>What do you mean by that? Especially, what the word epoch means? It's one of the hyperparameters that we define in the dictionary? I think maybe I miss some details of how Talos work.

Training a deep learning model is expressed as "epochs" on one hand. The very simple version is that an epoch is an event when the entire dataset is passed through the neural net. Each epoch builds on the previous epoch, which explains why the first epoch tends to have weaker performance than the following ones. When this does not happen, there is usually something wrong with the model or data or parameters. In some cases the best performance takes place during early epochs, sometimes late epochs, and sometimes on the last epoch. When the last epoch performs the best, it usually means there is still room to improve performance. You can think of it like someone gives you a paper to read with 5 pages. You're going to read that paper 10 times to really make it clear for yourself. As you read it again and again, you keep building on your previous learning. Each of those times is an epoch. 

So if you have  ta.Scan(... `last_epoch_value=True` ...) then you will see in Talos results the result of the last epoch. If you have ta.Scan(... `last_epoch_value=False` ...) then you will see the peak epoch, which could have been very early in the training process, and in later epochs the result might have become worse. The model might have started to over-train etc.

**Comment by YUIWK** [comment]
Sorry for the late reply. Thanks for your clarification! I do know how a model is trained in deep learning. Your clarification helps me understand Talos better. I am running the experient again to see if I can reproduce results. 
As for the the reproducibility problem, besides ta.Deploy() is there another way to get the same results from Scan? Sometimes I want to not only the metric like val_loss but also the predicted value to make a comparison with the actual values. I think it's pretty important or what we got from Scan results is meaningless and unreliable.

**Comment by mikkokotila** [comment]
There is no reproducibility problem in Talos. Talos does not take away anything or add anything to your Keras model. So if you are able to reproduce results with Keras model, then you are able to reproduce results with Talos.

If you want to take things offline or to another system, then the only way you can do it is using `ta.Deploy()` which gives you full access to models from every permutation . Basically `ta.Restore()` gives you a replica of the `ta.Scan()` object with some minor constraints. You can explore this easily yourself by looking at the namespaces/attributes for `ta.Scan()` and `ta.Restore()` objects. Talos namespaces are 100% clean i.e. there is nothing extra in the namespace so anything you see there is worth checking out to understand what is possible. 

Once you do that, you will also learn about options you have with `ta.Scan()` object, which seems to be your other question. You can also check `ta.Evaluate()` and `ta.Predict()`.

**Comment by mikkokotila** [comment]
> Sometimes I want to know not only the metric like val_loss but also the predicted value to make a comparison with the actual values. I think it's pretty important or what we got from Scan results is meaningless and unreliable.

If you are unclear with basic functionality of Talos, you will find the [README](https://github.com/autonomio/talos), [EXAMPLES](https://github.com/autonomio/talos#examples), and [DOCS](https://autonomio.github.io/docs_talos/#introduction) useful. Improvement suggestions to that end are most welcome.
